config/menu.php

```diff
@@ -0,0 +1,113 @@
+use Symfony\Component\DependencyInjection\Argument\ServiceLocatorArgument;
+use Symfony\Component\DependencyInjection\Argument\TaggedIteratorArgument;
+
+return static function (ContainerConfigurator $configurator): void {
+    $configurator->parameters()->set('knp_menu.renderer.list.options', []);
+
+    $services = $configurator->services();
+
+    $services
+        ->set('knp_menu.factory', MenuFactory::class)
+        ->public()
+    ;
+
+    $services
+        ->set('knp_menu.factory_extension.routing', RoutingExtension::class)
+        ->arg(0, service('router'))
+        ->tag('knp_menu.factory_extension')
+    ;
+
+    $services
+        ->set('knp_menu.helper', Helper::class)
+        ->args([
+            service('knp_menu.renderer_provider'),
+            service('knp_menu.menu_provider'),
+            service('knp_menu.manipulator'),
+            service('knp_menu.matcher'),
+        ])
+    ;
+
+    $services
+        ->set('knp_menu.matcher', Matcher::class)
+        ->public()
+        ->arg(0, new TaggedIteratorArgument('knp_menu.voter'))
+        ->tag('kernel.reset', ['method' => 'clear'])
+    ;
+
+    $services
+        ->set(MenuProviderInterface::class, service('knp_menu.menu_provider'))
+    ;
+
+    $services
+        ->set('knp_menu.menu_provider.chain', ChainProvider::class)
+        ->arg(0, new TaggedIteratorArgument('knp_menu.provider'))
+    ;
+
+    $services
+        ->set('knp_menu.menu_provider.lazy', LazyProvider::class)
+        ->arg(0, [])
+        ->tag('knp_menu.provider')
+    ;
+
+    $services
+        ->set('knp_menu.menu_provider.builder_alias', BuilderAliasProvider::class)
+        ->args([
+            service('kernel'),
+            service('service_container'),
+            service('knp_menu.factory'),
+        ])
+    ;
+
+    $services
+        ->set('knp_menu.renderer_provider', PsrProvider::class)
+        ->args([
+            new ServiceLocatorArgument(new TaggedIteratorArgument('knp_menu.renderer', null, 'alias')),
+            '%knp_menu.default_renderer%',
+        ]);
+
+    $services
+        ->set('knp_menu.renderer.list', ListRenderer::class)
+        ->tag('knp_menu.renderer', ['alias' => 'list'])
+        ->args([
+            service('knp_menu.matcher'),
+            '%knp_menu.renderer.list.options%',
+            '%kernel.charset%',
+        ]);
+
+    $services
+        ->set('knp_menu.voter.callback', CallbackVoter::class)
+        ->tag('knp_menu.voter');
+
+    $services
+        ->set('knp_menu.voter.router', RouteVoter::class)
+        ->arg(0, service('request_stack'))
+        ->tag('knp_menu.voter');
+
+    $services
+        ->set('knp_menu.manipulator', MenuManipulator::class)
+    ;
+
+    // Autowiring aliases
+    $services->set(FactoryInterface::class, service('knp_menu.factory'));
+    $services->set(MatcherInterface::class, service('knp_menu.matcher'));
+    $services->set(MenuManipulator::class, service('knp_menu.manipulator'));
```

Comment:
```suggestion
    $services->alias(FactoryInterface::class, 'knp_menu.factory');
    $services->alias(MatcherInterface::class, 'knp_menu.matcher');
    $services->alias(MenuManipulator::class, 'knp_menu.manipulator');
```

Reply:
Thanks for the suggestions. The main service now seems to be properly defined, but as soon as I try to use any menu I get a missing reference for the twig service. The service is defined in my project, but (quoting) the container inside "Symfony\Component\DependencyInjection\Argument\ServiceLocator" is a smaller service locator that only knows about the "knp_menu.renderer.list" and "knp_menu.renderer.twig" services."

Reply:
I'm testing this branch on an actual project, where Twig is present.

Reply:
It's the service locator that's broken. See my comment on the menu.php file.